### PR TITLE
Implement registry server for declarative config

### PR DESCRIPTION
**Pull request by joelanford**
**Description of the change:**
This PR builds on #617 by adding support for serving the GRPC API from a declarative config format.

A new `opm serve` command will read the config directory, build an internal model of all packages/channels/bundles, and then serve the existing GRPC query interface from that model. 

**Motivation for the change:**
See https://github.com/operator-framework/enhancements/blob/master/enhancements/declarative-index-config.md

**Reviewer Checklist**
- [x] Implementation matches the proposed design, or proposal is updated to match implementation
- [x] Sufficient unit test coverage 
- [x] Sufficient end-to-end test coverage
- [ ] Docs updated or added to `/docs` 
- [ ] Commit messages sensible and descriptive

### CODE REVIEW

File: cmd/opm/registry/serve.go
Comment (njhale) [inline cmd/opm/registry/serve.go:55]: I think a much better UX would be that the `registry` command takes a path as a required operand. That path can either be to a single database file, or to a config directory. Having multiple options when there can only be one source for serve is confusing IMO.

File: pkg/declcfg/declcfg.go
Comment (njhale) [inline pkg/declcfg/declcfg.go:213]: isn't this verification useful for all potential models (not just ones generated from declarative config?)

File: pkg/model/model.go
Comment (njhale) [inline pkg/model/model.go:133]: Should the ownership of the IR->$format belong to the specific backend? e.g. in the `api` package?

File: cmd/opm/registry/serve.go
Comment (joelanford) [reply cmd/opm/registry/serve.go]: Updated to use a positional argument, deprecate the `--database` flag, and use file type detection to load the appropriate registry source.

File: pkg/declcfg/declcfg.go
Comment (joelanford) [reply pkg/declcfg/declcfg.go]: Agreed. Latest commit adds (mostly) comprehensive validation to the model package.

File: pkg/model/model.go
Comment (joelanford) [reply pkg/model/model.go]: Done. I moved the conversion functions related to the `api` types to the `api` package. 

With that change, I was forced to move the `model.Querier` struct to another package to avoid an import cycle between `model` and `api`.  For now its in `registry`, but I'm not sure if that's the right place either.

File: cmd/opm/registry/serve.go
Comment (timflannagan) [inline cmd/opm/registry/serve.go:44]:
```suggestion
		Long:  `serve an operator-registry source that is queryable using grpc`,
```

File: cmd/opm/registry/serve.go
Comment (timflannagan) [inline cmd/opm/registry/serve.go:34]: Real minor nit: Does this need to be a global variable for the registry package? It looks like most of the existing code attempts to get the value of this flag through `cmd.Flags().Get...`.

File: cmd/opm/registry/serve.go
Comment (timflannagan) [inline cmd/opm/registry/serve.go:110]: Another minor nit: this likely can be reduced to an if/else given you return early in while stat-ing the source directory.

File: cmd/opm/registry/serve.go
Comment (joelanford) [reply cmd/opm/registry/serve.go:34]: Yeah I should probably either use hardcoded `"database"` everywhere or change other `cmd.Flags().Get` calls to use a const. Whenever I see a string used this way in more than one place, my default is to change it to a const, to avoid typo mistakes and make use of IDE auto-completion.

As an aside, using `cmd.Flags().Get()` is much more cumbersome than just declaring a flags struct and populating fields in the struct, like this:

```go
type serve struct {
	debug bool
	database string
	port string
	...
}

func (s serve) RunE(cmd *cobra.Command, args []string) error {
	// populate from serveFunc, replace `cmd.Flags().Get("<field>")` with s.<field>
}

func newRegistryServeCmd() *cobra.Command {
	s := serve{}
	rootCmd := &cobra.Command{
		...
		RunE: s.RunE,
	}
	cmd.Flags().BoolVar(&s.debug, "debug", ...)
	cmd.Flags().StringVarP(&s.database, "database", ...)
	cmd.Flags().StringVarP(&s.port, "port", ...)
	...
}
```

File: cmd/opm/registry/serve.go
Comment (joelanford) [reply cmd/opm/registry/serve.go:110]: Now I'm re-reading that section of code and I'm trying to figure out if I mistakenly used `err` when I should have used `sErr` if `os.Stat` returns an error.

Comment (joelanford) [reply cmd/opm/registry/serve.go:108]: ^ Fixed

File: cmd/opm/registry/serve.go
Comment (joelanford) [reply cmd/opm/registry/serve.go]: For now, reverted to hard-coded strings to be consistent with how other flags work.

In a separate PR, I may see how hard it would be to migrate the existing cobra usage over to structs with flag fields to avoid this issue entirely.

File: pkg/api/properties.go
Comment (joelanford) [inline pkg/api/properties.go:40]: Some of this code is more or less duplicated in the `pkg/api` package. It initially seemed like a good idea to scope the property parsing to the component that needs it and then ONLY parse the properties required by the component.

But some properties have two or more components that are interested in them, so it seems reasonable to centralize property definitions and parsing code as well. Thoughts?

File: pkg/declcfg/helpers_test.go
Comment (dinhxuanvu) [inline pkg/declcfg/helpers_test.go:197]: Where is buildMandalorianPkgModel()?

File: pkg/declcfg/bundle.go
Comment (dinhxuanvu) [inline pkg/declcfg/bundle.go:7]: I'm gonna be honest here by saying that I don't see the value of this Bundle API given it is pretty much identical to Bundle on registry package. The diff I can see here is the registry.Bundle has more fields and this bundle has `RelatedImages`. I understand this is helpful for the validation but I wonder if we reinvent the wheel here because I can see we potentially add more fields and it will become exactly like registry.Bundle.

File: test/e2e/testdata/configs/etcd.json
Comment (dinhxuanvu) [inline test/e2e/testdata/configs/etcd.json:1]: Is this config valid or invalid? This json format looks outdated (pre-your-enhancement-PR).

File: pkg/declcfg/bundle.go
Comment (dinhxuanvu) [reply pkg/declcfg/bundle.go:7]: I recall our conversation about moving this to api repo. I think that makes more sense.

File: test/e2e/testdata/configs/etcd.json
Comment (joelanford) [reply test/e2e/testdata/configs/etcd.json:1]: Yeah that's a leftover unused file. I'll delete it.

File: pkg/declcfg/bundle.go
Comment (joelanford) [reply pkg/declcfg/bundle.go:7]: I disagree about re-using `registry.Bundle` in place of the declarative config `bundle` struct. `declcfg.bundle` is the Go specification of the `olm.bundle` schema. It is used to (un)marshal declarative config JSON. That should be its singular purpose.

Looking at `registry.Bundle`, I get the impression that its primary purpose is the representation of the _contents_ of the bundle image, used when we unpack the bundle image and need to understand what it contains.

IMO using one struct to represent both declarative config's `olm.bundle` schema AND the contents of bundle described by that schema will cause unnecessary complexity. For instance, if we need to add a new field to `registry.Bundle` to capture some detail about the contents of the bundle, we also then need to understand how that might impact that struct's use as the specification for the `olm.bundle` spec.

Comment (joelanford) [reply pkg/declcfg/bundle.go:7]: I theory, I would agree that `operator-framework/api` is a logical place for `pkg/declcfg`. I think my suggestion for now would be to keep it all encapsulated in `operator-registry` so that we can iterate on it (if necessary) without involving another repository.

We can always move it `operator-framework/api` once it stabilizes. To prevent others from making use of it for now, we could put it in an `internal` tree to make it available only to other `operator-registry` packages.

Related side question: does `operator-registry` make any Go API guarantees? Given that there are v1.x tags, I would assume that we're not allowed to remove or break any exported APIs?

Comment (dinhxuanvu) [reply pkg/declcfg/bundle.go:7]: > I disagree about re-using `registry.Bundle` in place of the declarative config `bundle` struct. `declcfg.bundle` is the Go specification of the `olm.bundle` schema. It is used to (un)marshal declarative config JSON. That should be its singular purpose.
> 
> Looking at `registry.Bundle`, I get the impression that its primary purpose is the representation of the _contents_ of the bundle image, used when we unpack the bundle image and need to understand what it contains.
> 
> IMO using one struct to represent both declarative config's `olm.bundle` schema AND the contents of bundle described by that schema will cause unnecessary complexity. For instance, if we need to add a new field to `registry.Bundle` to capture some detail about the contents of the bundle, we also then need to understand how that might impact that struct's use as the specification for the `olm.bundle` spec.

1. I understand your intention to use this `declcfg.bundle` as the representation of `olm.bundle` JSON struct in the config and you are correct regarding `registry.bundle` is the representation of the bundle and its content. However, you blur the line between the two when you include `CsvJSON` and `Objects` in the `declcfg.bundle`. Those two fields are for bundle content so you are duplicating the `registry.bundle` purpose. I can see the comment you have on those two fields about preserving the compatibility with gRPC which actually colliding with another API we have also named Bundle (`api.Bundle` which is for gRPC API). As you can see, we end up with 3 different `something.Bundle` that represent essentially the same thing. I'm not saying this `declcfg.bundle` is a bad thing given we can't really change `api.Bundle` because that would break bundle gRPC API and `registry.Bundle` is kind of a mess due to legacy reason (we don't want to remove fields and keep adding more as we introduced `property` and `dependency` for resolution). So I can see this API perhaps is a necessary step to move forward with config work without dealing with existing API. However, it seems to become a troubling pattern that we have 3 APIs with the same name floating around in the same repo that represents the same thing perhaps in slightly different shapes or forms. Imagining someone looking our registry codebase and wonder why there are 3 different bundles here and frankly I wonder if they even notice these are 3 different APIs or they just assume it is the same thing.
2. We can decide where to keep this bundle API at a later time. I don't feel strongly about where it should belong though as I mentioned above, perhaps it should be in api for the sake of not having multiple bundle APIs in the same repo. Plus, it depends on who we think who is dictating the config format. I often think personally api repo is where OLM and registry have in common and we have everything there so they can share and reuse. In term of config, OLM doesn't use it directly so the config API is actually being dictated by registry which means it can stay here. There is nothing wrong with folks using API from registry repo directly.
3. I don't know if we guarantee the backward compatibility for any API registry except the gRPC API for bundle. The rest is pretty much for internal use. No one is using the APIs in registry directly as far as I know.

Comment (joelanford) [reply pkg/declcfg/bundle.go:7]: 1. Pretty much agree with everything you said. Perhaps we could beef up the GoDoc comments for each `Bundle` type to try to clarify where/why they're used?
2. I think we need to consider `operator-sdk` in the list of clients of `operator-framework/api`. I haven't thought about it _too_ much, but I could imagine `operator-sdk` needing to manipulate declarative configs at some point. Either way, my vote would be to leave it here for now as an API that can only be used internally by `operator-registry` components.

    Based on this, I'm thinking I'll move `declcfg` and `model` into an `internal` package so that other projects will be prevented from importing it. Then if/when there's a need for others to consume it, we can have this conversation again. And by then we'll have a concrete motivation to help us decide. WDYT?
3. This point is going off on a tangent, but I know [SDK imports `operator-registry`](https://github.com/operator-framework/operator-sdk/blob/b3547bcacaca9e10c7225ced6ebd6210067d7fc6/go.mod#L16), so there's at least one consumer of "internal" APIs. We should move everything that's actually internal to `internal` directory trees to prevent other projects from using them.

File: cmd/opm/registry/serve.go
Comment (joelanford) [inline cmd/opm/registry/serve.go:193]: Do we want to serve sqlite databases from the model, or should we just return the `dbStore` directly (which would mean continuing to serve sqlite from the sqlite querier)?

There's some risk involved in serving sqlite from the model because that's a change from current code paths that has not been thoroughly tested.

I suppose if we started getting bugs against this, it would be simple to revert back to serving directly from sqlite.

Comment (joelanford) [reply cmd/opm/registry/serve.go]: Plan discussed offline: by default keep serving sqlite databases as we always have.

If we want to load the sqlite db into memory and serve it from the model, we can do that really easily in a follow-up PR (probably gated by a feature flag).

File: pkg/declcfg/model_to_declcfg.go
Comment (anik120) [inline pkg/declcfg/model_to_declcfg.go:56]: If these are not set, for eg if CsvJSON and Objects for `chb` is not set during creating a bundle blob, won't this lead to a nil pointer dereferencing?

Comment (joelanford) [reply pkg/declcfg/model_to_declcfg.go:56]: This particular spot is okay as long as `chb` is non-nil. Which will always be the case unless there's a bug in our code that converts any of the representations to the model (or someone else is importing and using `model.Model` directly).

If `chb.CsvJSON` and/or `chb.Objects` is nil (definitely possible and happens frequently), then `declcfg` code should handle that. I'll check to see if the `Load` and `Write` function tests include cases for this (I think they do, but can't hurt to look again).

Comment (anik120) [reply pkg/declcfg/model_to_declcfg.go:56]: @joelanford here's an annotations.yaml file for the bundle quay.io/operatorhubio/etcd:v0.6.1

```
annotations:
  operators.operatorframework.io.bundle.channel.default.v1: ""
  operators.operatorframework.io.bundle.channels.v1: alpha
  operators.operatorframework.io.bundle.manifests.v1: manifests/
  operators.operatorframework.io.bundle.mediatype.v1: registry+v1
  operators.operatorframework.io.bundle.metadata.v1: metadata/
  operators.operatorframework.io.bundle.package.v1: etcd
```

The default channel is not set here, which means when we're creating the model from that bundle, the `model.Package.DefaultChannel` is nil. 
Since we're accessing `mpkg.DefaultChannel.Name` here directly, it'll run into a nil pointer dereferencing error

Comment (joelanford) [reply pkg/declcfg/model_to_declcfg.go:56]: If the model, has been [validated](https://github.com/operator-framework/operator-registry/pull/601/files#diff-79ca10368c383e5ed5c7d8885fe1f4c2758b5ae4afd3feaf629e3a348b69b054R57-R59), `mpkg.DefaultChannel` won't be nil.

And (at least for operator-registry's uses of the model package) the conversion functions run the validations.
- [Declarative config to model](https://github.com/operator-framework/operator-registry/pull/601/files#diff-345bb90148fa16097184a5f10d48e80bf0cc6f8bfa5b6c7054b41070a50235aeR84-R86)
- [Sqlite to model](https://github.com/operator-framework/operator-registry/pull/601/files#diff-806b9455cc4618f4bd3e89d89ab1cb18499ec2e67f21b2914f8e2b69b057dc7eR28-R30)

Nonetheless, you're right that this code could result in a panic if used in other places or ways.  I think there are two possible solutions:
1. Add a `nil` check here for `mpkg.DefaultChannel` and if `nil`, use an empty string.
2. Run the full model validation at the beginning of the function, which would guarantee that `mpkg.DefaultChannel` would be non-nil.

With option 1, we're _kind of_ saying, that it's okay for this field to be unset, and as far as I know, we currently enforce that every package has a default channel (correct me if I'm wrong).

With option 2, we might be enforcing validations that are too strict.

As an aside, I think we should relax the model validations down to enforce only what the GRPC API requires.

File: pkg/registry/query.go
Comment (benluddy) [inline pkg/registry/query.go:138]: This one appears to still used by catalog-operator:

https://github.com/operator-framework/operator-lifecycle-manager/blob/master/pkg/controller/operators/catalog/operator.go#L1011
https://github.com/operator-framework/operator-lifecycle-manager/blob/d9ff24161ab4f7d4d323afb662a56f453791329c/pkg/controller/registry/resolver/querier.go#L143
https://github.com/operator-framework/operator-registry/blob/a1fc9d0f1d2e04c14229801605ed727cd80e956e/pkg/client/client.go#L74

Comment (joelanford) [reply pkg/registry/query.go:138]: I'll double check, but I think there's some dead code in OLM, and this might be one of those cases.

Comment (joelanford) [reply pkg/registry/query.go:138]: Oh now I see you already did the leg work. Thanks!

In that case, I'll cross-reference with the sqlite implementation of this function and see if its deterministic (and if so, I'll use the same algorithm).

Comment (joelanford) [reply pkg/registry/query.go:138]: The sqlite implementation looks non-deterministic to me since there's no ORDER BY in the query.

The model implementation in this PR is ALSO non-deterministic because it iterates over a map and chooses the first bundle it finds that meets the criteria. Since map iteration is non-deterministic, the returned bundle is non-deterministic.

I'll remove this TODO comment, and add some comments in the code that call this out.

Comment (joelanford) [reply pkg/registry/query.go]: Done!

File: pkg/declcfg/helpers_test.go
Comment (joelanford) [reply pkg/declcfg/helpers_test.go:206]: It doesn't exist. Did I miss something?

File: pkg/declcfg/model_to_declcfg.go
Comment (joelanford) [reply pkg/declcfg/model_to_declcfg.go:60]: Fixed the `nil` pointer access issue with `model.Package.DefaultChannel.Name`.

File: pkg/property/property.go
Comment (dinhxuanvu) [inline pkg/property/property.go:233]: Can we comment on what are the criteria we are using here for backward compatibility?

File: pkg/model/model.go
Comment (dinhxuanvu) [inline pkg/model/model.go:130]: Answer to your question in the `TODO`. We currently use this ([link](https://github.com/operator-framework/operator-lifecycle-manager/blob/master/pkg/controller/registry/resolver/resolver.go#L591)) to sort bundles in a channel (which effectively finds the head of channel). You should consult with that code and perhaps we should align them.

Comment (joelanford) [reply pkg/model/model.go:130]: I think that's doing the same thing my code is doing. It's just that my code is not concerned with figuring out the sorted order of the entire channel.

Comment (dinhxuanvu) [reply pkg/model/model.go:130]: Then I think we should be fine with this.

File: pkg/property/property.go
Comment (joelanford) [reply pkg/property/property.go:233]: This is a bit of a relic function from 10s of commits ago and as you can see it's much shorter now. I'll revisit this to see if its necessary to keep the last remaining check separated out in this function.

File: pkg/model/model.go
Comment (joelanford) [reply pkg/model/model.go:130]: @ecordell Did mention that there are index maintainer use cases for potential multi-head channels, so I'll leave this comment in place, but update it to link to the SAT solver implementation and mention Evan's potential use case.

**Review by dinhxuanvu — COMMENTED** [review]
This PR needs to rebase with Joe's base PR because I still see code from the query which should be on Joe's serve PR only.
Plus, we should review the part that belongs to Joe's base PR on that PR, not here to avoid polluting the review process on the `add` part.

**Comment by joelanford** [comment]
@dinhxuanvu rebased!

**Comment by joelanford** [comment]
No idea what's going on here: https://github.com/operator-framework/operator-registry/pull/601/checks?check_run_id=2246459783#step:5:161

I can't get this to happen locally. I'm using go 1.16.2 and I'm running the same steps from the sanity test workflow.
```
$ go install golang.org/x/tools/cmd/goimports@latest
go: downloading golang.org/x/tools v0.1.0
go: downloading golang.org/x/sys v0.0.0-20210119212857-b64e53b001e4

$ make vendor && make lint && git diff --exit-code
GOFLAGS="-mod=vendor" go mod tidy
GOFLAGS="-mod=vendor" go mod vendor
GOFLAGS="-mod=vendor" go mod verify
all modules verified
find . -name '*.go' -not -path "./vendor/*" | xargs goimports -w
```

File: pkg/api/api_to_model.go
Comment (joelanford) [inline pkg/api/api_to_model.go:1]: Code that converts the GRPC API and sqlite DBs into the model isn't strictly required for `opm serve`. However it's used in `pkg/server/server_test.go` to test that both of the following scenarios result in the equivalent GRPC responses:
1. sqlite -> GRPC
2. sqlite -> model -> GRPC

File: pkg/registry/query.go
Comment (dinhxuanvu) [inline pkg/registry/query.go:254]: I wonder if it is beneficial to convert this func and `bundleReplaces` to public funcs on `model` packages.
For example:
`func(b model.Bundle) Replaces(name string) bool {}`
Just a thought.

File: pkg/api/model_to_api.go
Comment (dinhxuanvu) [reply pkg/api/model_to_api.go:26]: I do notice that the `parseProperties` func does validate this scenario so it won't happen.

**Review by dinhxuanvu — COMMENTED** [review]
Looking good. Just a few comments.

File: pkg/api/model_to_api.go
Comment (joelanford) [reply pkg/api/model_to_api.go:37]: The main reason for the difference is because `api.Bundle` has a `SkipRange` field that allows either 0 or 1 skip ranges. When converting to the `api.Bundle`, we have to verify that no more than one skip range is defined.

The model itself doesn't have this constraint built-in, which is why the declcfg_to_model function does not include this check.

File: pkg/registry/query.go
Comment (joelanford) [reply pkg/registry/query.go:254]: At one point, I think I had a version of the model package that had these kind of functions. Let me see if I can work out why I changed it.

Comment (joelanford) [reply pkg/registry/query.go:254]: Ahh. I think I did it this way because I wanted to keep property the model out of the business of parsing and interpreting properties, which would be necessary to answer these types of questions.

Comment (joelanford) [reply pkg/registry/query.go:254]: We _could_ put a `DoesReplace` function in model since `model.Bundle` has a single `Replaces` field in the struct. But the provided GVKs are all stored in opaque properties in the model.

My instinct is to leave the model pretty minimal. If there's a future where multiple packages have to implement these same sort of functions against the model, we could revisit.

Comment (dinhxuanvu) [reply pkg/registry/query.go:254]: I see. I agree that this func `doesModelBundleProvide` isn't necessarily beneficial to have on the model because it does require some conversion. The replace one is more straightforward to have on model. Either way, it is not a big deal. I just thought about reorganizing the codebase a bit but it is not a concern at the moment. Move on.

File: pkg/api/model_to_api.go
Comment (dinhxuanvu) [reply pkg/api/model_to_api.go:37]: I'm a bit confused. What do you mean by `allows either 0 or 1 skip ranges`? skipRange is just a string in `api.Bundle`. You mean with `model.Bundle`, they can have multiple `skipRange` properties?

Comment (joelanford) [reply pkg/api/model_to_api.go:37]: Correct.

File: pkg/sqlite/conversion.go
Comment (anik120) [inline pkg/sqlite/conversion.go:1]: Are we converting the sqllite db to model before serving the db anymore now that we've decided serving with configs is a separate command altogether? 

I couldn't find a place where we're doing that in this PR, maybe I'm missing something?

File: pkg/registry/query.go
Comment (anik120) [inline pkg/registry/query.go:24]: Are we testing these new implementations against the existing tests somewhere?

**Review by anik120 — COMMENTED** [review]
Looking really good so far @joelanford, had a couple of questions.

File: pkg/sqlite/conversion.go
Comment (joelanford) [reply pkg/sqlite/conversion.go:1]: You're correct. The `opm registry serve` command is completely unchanged, so sqlite DBs will be served directly (without converting to the model).

The reason this file is present in this PR is to enable the test in `./pkg/server/server_test.go`. In that test, we serve the DB directly, and we convert the DB to the model and serve that via the model-based GRPC server. This helps us guarantee that the sqlite GRPC server and the model-based GRPC server function the same against the same underlying data.

File: pkg/registry/query.go
Comment (joelanford) [reply pkg/registry/query.go:24]: Yes, `pkg/server/server_test.go`.

File: pkg/api/model_to_api.go
Comment (dinhxuanvu) [reply pkg/api/model_to_api.go:37]: Given this restriction on skipRange is universal, then I don't see the reason why we don't impose this restriction on `declcfg` side as well.

Comment (joelanford) [reply pkg/api/model_to_api.go:37]: > this restriction on skipRange is universal

I think that's the current state of the world with the current resolver and the current GRPC API. But that might not always necessarily be the case when we've updated the resolver and gotten rid of the GRPC server.

Comment (dinhxuanvu) [reply pkg/api/model_to_api.go:37]: Sure. I can see that you want to model side to be more generic and not discriminating against this. I can see a future where the api.Bundle no longer needs a specific field for `skipRange` (or no longer exists in the new non-grpc API we use) but I don't see a future where we allow multiple `skipRange, right? However, the fact that we use the model to validate config. If the config is considered valid even with multiple `skipRange`, then it will fail the serve part. That is not a good UX.

File: pkg/registry/query.go
Comment (anik120) [reply pkg/registry/query.go:24]: Okay I see it (for some reason github decided to contract that file while I was reviewing it so it escaped my eye 🤦🏽 )

Comment (joelanford) [reply pkg/registry/query.go:24]: Yeah GitHub has some weird behaviors like that. Super annoying.

**Comment by anik120** [comment]
/lgtm 
if @dinhxuanvu's remaining concern is addressed it looks good to me.

File: pkg/api/model_to_api.go
Comment (joelanford) [reply pkg/api/model_to_api.go:37]: Chatted offline with @dinhxuanvu. We'll handle this concern in #618

Comment (dinhxuanvu) [reply pkg/api/model_to_api.go:37]: Joe and I agree that we should have this "validation" on the model side so the validate and serve can be in sync in term of pass/fail on the same scenario which in case is multiple skipRange. We will keep things as they are in Joe's PR and will address this on my PR (which should be merged after Joe's).

**Review by dinhxuanvu — APPROVED** [review]
/lgtm

**Pull request merged**
